presage/src/manager.rs

```diff
@@ -1126,30 +1120,54 @@ impl<C: Store> Manager<C, Registered> {
         })
     }
 
+    /// Return a clone of a cached _unidentified_ push service.
+    fn unidentified_push_service(&self) -> Result<HyperPushService, Error<C::Error>> {
+        self.state.unidentified_push_service.get(|| {
+            let service_configuration: ServiceConfiguration = self.state.signal_servers.into();
+            Ok(HyperPushService::new(
+                service_configuration,
+                None,
+                crate::USER_AGENT.to_string(),
+            ))
+        })
+    }
+
+    async fn identified_websocket(&self) -> Result<SignalWebSocket, Error<C::Error>> {
+        if let Some(ws) = self.state.identified_websocket.borrow().as_ref() {
+            return Ok(ws.clone());
+        }
+
+        let ws = self
+            .identified_push_service()?
+            .ws("/v1/websocket/", self.credentials()?, true)
+            .await?;
+        self.state.identified_websocket.replace(Some(ws.clone()));
+        Ok(ws)
+    }
+
+    async fn unidentified_websocket(&self) -> Result<SignalWebSocket, Error<C::Error>> {
+        if let Some(ws) = self.state.unidentified_websocket.borrow().as_ref() {
```

Comment:
I had a similar code which caches the unidentified websocket. The question is: what happens if connection is lost to it? Then we have a cached broken websocket. Right?

Reply:
For now, this is unhandled, which is really bad and the reason why I haven't merged this PR yet.